scylla-proxy/src/proxy.rs

```diff
@@ -681,22 +683,27 @@ impl Doorkeeper {
         real_addr: SocketAddr,
         shard_awareness: ShardAwareness,
     ) -> Result<TcpStream, DoorkeeperError> {
-        let cluster_stream = if shard_awareness.is_aware() {
-            let shards = match self.shards_number {
-                None => {
-                    let temporary_stream = TcpStream::connect(real_addr)
-                        .await
-                        .map_err(|err| DoorkeeperError::NodeConnectionAttempt(real_addr, err))?;
-                    let shards = self
-                        .obtain_shards_number(temporary_stream, real_addr)
-                        .await?;
-                    self.shards_number = Some(shards);
-
-                    shards
-                }
-                Some(shards) => shards,
-            };
+        let shards = if shard_awareness.is_aware() {
+            if let shards @ Some(_) = self.shards_number {
+                shards
+            } else {
+                let temporary_stream = TcpStream::connect(real_addr)
+                    .await
+                    .map_err(|err| DoorkeeperError::NodeConnectionAttempt(real_addr, err))?;
+                let shards = match self.obtain_shards_number(temporary_stream, real_addr).await {
+                    Ok(shards) => Some(shards),
+                    // If a node offers no sharding info, change proxy ShardAwareness to Unaware.
+                    Err(DoorkeeperError::ObtainingShardNumberNoShardInfo) => None,
```

Comment:
It's a bit strange that `obtain_shards_number` considers a situation where there is no shard info in the SUPPORTED response to be an error. This just indicates that the cluster behind the proxy is a Cassandra cluster (or a Scylla cluster which does not advertise its shard count - there is an obscure configuration option for this which nobody uses).

Reply:
Just ranting, it's not a big issue, it's fine to fix this later.

Reply:
It is an error at least in `obtain_shards_count`'s sense: after all, it has failed to obtain shards count.